pennylane/operation.py

```diff
@@ -468,15 +468,27 @@ def __init__(self, *params, wires=None, do_queue=True, id=None):
         if wires is None:
             raise ValueError(f"Must specify the wires that {self.name} acts on")
 
-        self._num_params = len(params)
-        # Check if the expected number of parameters coincides with the one received.
-        # This is always true for the default `Operator.num_params` property, but
-        # subclasses may overwrite it to define a fixed expected value.
-        if len(params) != self.num_params:
-            raise ValueError(
-                f"{self.name}: wrong number of parameters. "
-                f"{len(params)} parameters passed, {self.num_params} expected."
-            )
+        # Check whther the expected number of parameters (in case the Operation subclass defines it via the num_params property) coincides with the one received.
+        if hasattr(self.__class__, "num_params"):
+            # pylint: disable=no-member, access-member-before-definition
+            if (
+                isinstance(self.__class__, int)
+                and len(params) != self.__class__.num_params
+                or self.num_params != len(params)
+            ):
+                raise ValueError(
+                    f"{self.name}: wrong number of parameters. "
+                    f"{len(params)} parameters passed, {self.__class__.num_params} specified in class {self.__class__}."
+                )
+        else:
+            # If no num_params property was provided by the-sub class we set it
+            # here. This allows sub-classes to define num_params as a class
+            # property in case it is the same for all intsances (which is the case
+            # for almost all Operations and allows to write more advanced circuit
+            # synthesis code that needs to know the number of paramters, e.g.,
+            # for instanciating the Operation with the correct number of
+            # parameters when given the class only)
```

Comment:
```suggestion
            # The number of parameters is not a static class property and 
            # inferred from the number of positional arguments received.
```

I am a fan of long comments, but yours makes claims that may not hold in future (i.e. if our template library outgrows the elementary gates) and explains very specific use cases. Do you mind if we keep it short and just state what is done?

Reply:
This was more to justify the RP. Feel free to shorten as you see fit.